audit: log user IP address

taylorsilva opened the pull request:
## What does this PR accomplish?
Bug Fix | **Feature** | Documentation

From concourse/concourse#6456

## Changes proposed by this PR:
Has the auditor also log a users IP address. Will log based on available headers (X-REAL-IP and X-FORWARDED-FOR) and fall back to `RemoteAddr` if no header is available.


## Release Note

* The audit logs now log a user's IP address

## Contributor Checklist
- [x] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [x] [Signed] all commits
- [x] Added tests (Unit and/or Integration)
- [x] Updated [Documentation]
- [x] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

taylorsilva committed:
audit: log user IP address

A security audit from a customer noticed that IP addresses are not
logged. Having this info logged helps investigations where a users
account was compromised and finding out which workstation it was used
from.

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

aoldershaw commented on atc/auditor/auditor.go:
```diff
@@ -164,6 +164,6 @@ func (a *auditor) ValidateAction(action string) bool {
 	err := r.ParseForm()
 	if err == nil && a.ValidateAction(action) {
-		a.logger.Info("audit", lager.Data{"action": action, "user": userName, "parameters": r.Form})
+		a.logger.Info("audit", lager.Data{"action": action, "user": userName, "ip": r.RemoteAddr, "parameters": r.Form})
```
Small note: `r.RemoteAddr` includes both the IP address *and* the port (which is going to be some random higher-order ephemeral port) - I wonder if we want to strip that out? Or possibly include it as a separate field? Or just rename `ip` to `addr`?

We could use `net.SplitHostPort` if we want to extract the ip in isolation

taylorsilva replied:
Seems that getting a users real ip can be tricky... https://golangbyexample.com/golang-ip-address-http-request/

I'm going to change it to check those headers in the order of `X-REAL-IP`, `X-FORWARDED-FOR`, and then RemoteAddr, parsing it if needed.

aoldershaw replied:
Hm, does that mean you'd be able to spoof the IP that gets logged by providing one of those headers? Not sure if nginx or other reverse proxies would overwrite this header, but if not (or you aren't using a reverse proxy), it'd be easy to lie about your IP address (in which case, is there much value in including it in the audit log?)

taylorsilva committed:
Use X-* headers for getting a client's IP address

Getting a client's IP address is a bit harder than initially thought.
There are two headers that one should check before falling back to the
RemoteAddr available in the request (which is guaranteed to always be
set btw).

The auditor will now check for X-REAL-IP and X-FORWARDED-FOR headers and
log those, in that order, as the client's IP. This is not foolproof and
is easily spoofed by a client.

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

aoldershaw commented:
I'm a little torn - the feature request was for security auditing, but using `X-Real-Ip` and `X-Forwarded-For` mean that the IP address can easily be spoofed by a malicious user, making the audit log unreliable. That said, it's likely pretty common to have a reverse proxy in front of Concourse, making `r.RemoteAddr` useless. However, in that case, maybe it's just the responsibility of the reverse proxy to keep an audit log of requests (something they do, anyway) - it won't be the same granularity as Concourse's audit log since it doesn't know about actions in Concourse, but at least it can't be spoofed and includes the requested path (which is enough to deduce the action).

Perhaps an alternative (higher-effort) solution is to allow logging a custom header (configured via flags). That way, users could choose to log `X-Real-Ip` or `X-Forwarded-For` (opting-in to spoofing attacks), or they could log something else entirely. For instance, the reverse proxy could set a unique request ID in `X-Request-ID`, and Concourse could log that - to find the actual IP, they could consult their reverse proxy's logs for that request ID.

taylorsilva committed:
auditor: add custom flag for logging clients' IP

A user can configure their reverse proxy to set a custom header that
contains a users IP address if they're worried about client's spoofing
the X-* headers.

IP's are also parsed to ensure it's actually an IP being logged and not
some garbage string.

Signed-off-by: Taylor Silva <tsilva@pivotal.io>

taylorsilva commented:
> I'm a little torn - the feature request was for security auditing

I agree with this entirely, it's all I was thinking about as I was writing it and noted as much in my commit message

> This is not foolproof and is easily spoofed by a client.

Went ahead and added a config option for adding a custom header. It will still fallback to the other headers and remoteAddr.

aoldershaw commented on atc/atccmd/command.go:
```diff
@@ -214,15 +214,16 @@ type RunCommand struct {
+		EnableTeamAuditLog      bool   `long:"enable-team-auditing" description:"Enable auditing for all api requests connected to teams."`
+		EnableWorkerAuditLog    bool   `long:"enable-worker-auditing" description:"Enable auditing for all api requests connected to workers."`
+		EnableVolumeAuditLog    bool   `long:"enable-volume-auditing" description:"Enable auditing for all api requests connected to volumes."`
+		ClientIPHeader          string `long:"client-ip-header" description:"A header that a reverse proxy will set to the client's IP address. If not set then X-REAL-IP and X-FORWARD-FOR will be used in that order."`
```
I was thinking more along the lines of just adding some header(s), not necessarily for the IP address, that we just pass along as is and (don't try to infer any meaning from). In addition to logging the custom header(s), we'd log `r.RemoteAddr` as is.

So, if the operator configured logging the header `X-Forwarded-For`, then we'd log something like:

```
{"action": "...", "user": "...", "ip": "<remote proxy IP>", headers: {"X-Forwarded-For": "<local ip, possibly a chain, possibly spoofed>"}}
```

Or, if the operator configured logging the header `X-Request-Id`:

```
{"action": "...", "user": "...", "ip": "<remote proxy IP>", headers: {"X-Request-Id": "<some id that can be traced through the remote proxy logs>"}}
```

What do you think about that? I think always logging `r.RemoteAddr` is a good idea, since we can't possibly know what configuration of proxies is sitting in front of Concourse, if any. Allowing them to log any header(s) also seems like a good idea, since it allows you to log things like `X-Request-Id`, which can't be spoofed

cirocosta replied:
> Allowing them to log any header(s) also seems like a good idea, since it allows you to log things like X-Request-Id, which can't be spoofed (if you've configured your reverse proxy correctly).

and (just to add to the "it's hard to get a meaningful ip"), nowadays it's just sooooo easy to use any of those browsers w/ VPN support built-in, or even [tor](https://www.torproject.org/) (one can literally just go to a private tab on brave and enable it or download the tor browser)

aoldershaw replied:
@taylorsilva just pinging you about this - any thoughts about this approach?

taylorsilva commented:
I think the approaches mentioned make sense. At the same time, the scope on this is growing a bit much. I think it goes working towards a great final solution but I wasn't out to get to that point in this PR. I was looking to do a first pass. I'm gonna close this PR for now and make an issue from the discussion thread. I'll put it in the icebox.
